tests/flow/test_foreach.py

```diff
@@ -0,0 +1,482 @@
+                           CREATE (t:TEMP {v: i}))""")
+
+        # validate the creation
+        self.get_res_and_assertEquals("""MATCH (t:TEMP) RETURN t.v, count(t.v)
+                                         ORDER BY t.v""",
+                                      [[1, 2], [2, 2], [3, 2], [4, 2]])
+
+    # mid-evaluation failure (memory free'd appropriately)
+    def test07_midfail(self):
+        # clean db
+        self.env.flush()
+
+        try:
+            graph.query("FOREACH(i in [1, 2, 0, 3] | CREATE (n:N {v: 1/i}))")
+        except redis.exceptions.ResponseError as e:
+            self.env.assertIn("Division by zero", str(e))
+
+        # assure that no nodes are left (undo-log was applied correctly)
+        res = graph.query("MATCH (n) return count(n)")
+        self.env.assertEquals(res.result_set[0][0], 0)
+
+    # complicate things up
+    def test08_complex(self):
+        global graph
+
+        # ----------------------------------------------------------------------
+        # addressing node properties in the record sent to Foreach
+        # ----------------------------------------------------------------------
+
+        # create a single node
+        graph.query("CREATE (:N {v: 1, name: 'RAZ', li: [1, 2, 3, 4]})")
+
+        query = """
+        MATCH (n:N {v: 1, name: toUpper('raz')})
+        FOREACH(x in n.li |
+            CREATE (m:M {v: n.v, name: toLower(n.name), li: [x]})
+        )
+        """
+
+        res = graph.query(query)
+
+        # validate the query statistics
+        self.env.assertEquals(res.nodes_created, 4)
+        self.env.assertEquals(res.properties_set, 12)
+        self.env.assertEquals(res.nodes_deleted, 0)
+
+        # validate that the end state is correct
+        # there should be one node in the graph, with label N, 
+        # properties: {v: 2, name: 'RAZmon', li: [1] and 4 nodes with label M
+        # and properties: {v: 1, name: 'raz'} and li from 1 to 4
+        res = graph.query("MATCH (n:N) return n")
+        n = Node(label='N', properties={'v': 1, 'name': 'RAZ', 'li': [1, 2, 3, 4]})
+        self.env.assertEquals(res.result_set[0][0], n)
+        self.env.assertEquals(len(res.result_set), 1)
+        res = graph.query("MATCH (m:M) return m")
+        self.env.assertEquals(len(res.result_set), 4)
+        for i in range(len(res.result_set)):
+            m = Node(label= 'M', properties={'v': 1, 'name': 'raz', 'li': [i+1]})
+            self.env.assertEquals(res.result_set[i][0], m)
+
+        # clear nodes with label M
+        res = graph.query("MATCH (m:M) DELETE m")
+        self.env.assertEquals(res.nodes_deleted, 4)
+
+        # ----------------------------------------------------------------------
+        # triple embedding of a FOREACH clause, referencing node properties
+        # ----------------------------------------------------------------------
+
+        query = """
+        MATCH (n:N {v: 1, name: toUpper('raz'), li: [1, 2, 3, 4]})
```

Comment:
we can simplify this from: `MATCH (n:N {v: 1, name: toUpper('raz'), li: [1, 2, 3, 4]})`
to: `MATCH (n:N)`